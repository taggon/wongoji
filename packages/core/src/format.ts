import { parse, type Quote, type SingleQuote } from './parse';
import type { WongojiConfig } from './config';

const WONGOJI_WIDTH = 20;

/**
 * 주어진 원고지 형식에 맞추어 텍스트를 이차원 배열로 반환한다.
 * 1단계 배열은 원고지 한 줄을 의미하고, 2단계 배열은 줄의 한 칸을 의미한다.
 * @param text
 * @param config
 * @returns
 */
export function format(text: string, config: WongojiConfig = {}): string[][] {
  const doc = parse(text);

  return doc.nodes.reduce((lines, { type, nodes }) => {
    let currentLine: string[] = [];

    // 일단 새 줄을 추가하고 시작
    lines.push(currentLine);

    // 새 문단이나 인용문을 시작할 때는 한 칸 비우고 시작한다.
    currentLine.push(' ');

    for (const node of nodes) {
      switch (node.type) {
        case 'fulltext':
          let text = node.raw;
          while (currentLine.length + text.length > WONGOJI_WIDTH) {
            const first = text.slice(0, WONGOJI_WIDTH - currentLine.length);
            text = text.slice(WONGOJI_WIDTH - currentLine.length);
            currentLine.push(...first.split(''));
            lines.push(currentLine = []);
          }
          currentLine.push(...text.split(''));
          break;
        case 'halftext':
          for (let i=0; i < node.raw.length; i+=2) {
            const ch = node.raw.slice(i, i + 2);
            currentLine.push(ch);

            if (currentLine.length === WONGOJI_WIDTH) {
              lines.push(currentLine = []);
            }
          }
          break;
        case 'space':
          if (currentLine.length > 0) {
            currentLine.push(' ');
          }
          break;
        case 'punctuation':
          if (currentLine.length === 0 && lines.length > 1) {
            const secondLastLine = lines[lines.length - 2];
            const lastText = secondLastLine.pop();
            secondLastLine.push(...currentLine);
          } else {
            currentLine.push(node.raw);
          }
          break;
      }

      if (currentLine.length === 20) {
        lines.push(currentLine = []);
      }
    }

    return lines;
  }, [] as string[][]);;
}
