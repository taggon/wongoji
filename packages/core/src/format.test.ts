import { expect, describe, test } from 'bun:test';
import { format } from './format';

// const testCases = [
//   '안녕하세요. 반갑습니다.',
//   '대한민국은 Korea이다.',
//   '걸음도 해깝고 방울소리가 밤 벌판에 한층 청청하게 울렸다.\n달이 어지간히 기울어졌다.',
//   '"이걸 까셀라 부다!"하고 소리를 쳤다.',
//   // '대체 이게 웬 속인지(지금까지도 난 영문을 모른다) 아버질 혼내 주기는 제가 내래 놓고 이제 와서는 달려들며, "에그머니! 이 망할 게 아버지 죽이네!" 하고 내 귀를 뒤로 잡아당기며 마냥 우는 것이 아니냐.'
// ];

// testCases.forEach((text, index) => {
//   test(`Case #${index + 1}: ${text}`, () => {
//     expect(format(text)).toMatchSnapshot();
//   });
// });

describe('간단한 문장.', () => {
  let i = 1;

  test(`Case #${i++}: 안녕하세요. 반갑습니다.`, () => {
    expect(format('안녕하세요. 반갑습니다.')).toEqual([
      [' ', '안', '녕', '하', '세', '요', '.', '반', '갑', '습', '니', '다', '.'],
    ]);
  });

  test(`Case #${i++}: 대한민국은 Korea이다.`, () => {
    expect(format('대한민국은 Korea이다.')).toEqual([
      [' ', '대', '한', '민', '국', '은', ' ', 'K', 'or', 'ea', '이', '다', '.'],
    ]);
  });

  test(`Case #${i++}: 줄바꿈 처리`, () => {
    expect(format('걸음도 해깝고 방울소리가 밤 벌판에 한층 청청하게 울렸다.\n달이 어지간히 기울어졌다.')).toEqual([
      [' ', '걸', '음', '도', ' ', '해', '깝', '고', ' ', '방', '울', '소', '리', '가', ' ', '밤', ' ', '벌', '판', '에'],
      ['한', '층', ' ', '청', '청', '하', '게', ' ', '울', '렸', '다', '.'],
      [' ', '달', '이', ' ', '어', '지', '간', '히', ' ', '기', '울', '어', '졌', '다', '.'],
    ]);
  });
});

describe('따옴표 처리 웁션', () => {
  const text = '"이걸 까셀라 부다!"하고 소리를 쳤다.';

  test('스마트 인용 부호(useSmartQuote) 옵션이 true라면 따옴표와 홑따옴표를 유니코드 문자로 변경한다.', () => {
    expect(format(text, { useSmartQuote: true })).toEqual([
      [' ', '“', '이', '걸', ' ', '까', '셀', '라', ' ', '부', '다', '!', '”', '하', '고', ' ', '소', '리', '를', ' '],
      [' ','쳤', '다', '.'],
    ]);
  });

  test('인용문 항상 줄바꿈() 옵션이 true라면 인용문을 독립된 문단에 표시한다.', () => {
    expect(format(text, { useSmartQuote: true })).toEqual([
      [' ', '“', '이', '걸', ' ', '까', '셀', '라', ' ', '부', '다', '!', '”'],
      ['하', '고', ' ', '소', '리', '를', ' ','쳤', '다', '.'],
    ]);
  });
});
