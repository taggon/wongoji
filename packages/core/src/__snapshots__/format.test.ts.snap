// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Case: "이걸 까셀라 부다!"하고 소리를 쳤다. 인용문 항상 줄바꿈() 옵션이 true라면 인용문을 독립된 문단에 표시한다. 1`] = `
[
  [
    " ",
    "하",
    "고",
    " ",
    "소",
    "리",
    "를",
    " ",
    "쳤",
    "다",
    ".",
  ],
]
`;

exports[`Case #1: 안녕하세요. 반갑습니다. 1`] = `
[
  [
    " ",
    "안",
    "녕",
    "하",
    "세",
    "요",
    ".",
    "반",
    "갑",
    "습",
    "니",
    "다",
    ".",
  ],
]
`;

exports[`Case #2: 대한민국은 Korea이다. 1`] = `
[
  [
    " ",
    "대",
    "한",
    "민",
    "국",
    "은",
    " ",
    "K",
    "or",
    "ea",
    "이",
    "다",
    ".",
  ],
]
`;

exports[`Case #3: 걸음도 해깝고 방울소리가 밤 벌판에 한층 청청하게 울렸다.
달이 어지간히 기울어졌다. 1`] = `
[
  [
    " ",
    "걸",
    "음",
    "도",
    " ",
    "해",
    "깝",
    "고",
    " ",
    "방",
    "울",
    "소",
    "리",
    "가",
    " ",
    "밤",
    " ",
    "벌",
    "판",
    "에",
  ],
  [
    "한",
    "층",
    " ",
    "청",
    "청",
    "하",
    "게",
    " ",
    "울",
    "렸",
    "다",
    ".",
  ],
  [
    " ",
    "달",
    "이",
    " ",
    "어",
    "지",
    "간",
    "히",
    " ",
    "기",
    "울",
    "어",
    "졌",
    "다",
    ".",
  ],
]
`;

exports[`Case #4: "이걸 까셀라 부다!"하고 소리를 쳤다. 1`] = `
[
  [
    " ",
    "하",
    "고",
    " ",
    "소",
    "리",
    "를",
    " ",
    "쳤",
    "다",
    ".",
  ],
]
`;
